cylc/flow/pathutil.py

```diff
@@ -249,3 +283,57 @@ def get_next_rundir_number(run_path):
         return int(last_run_num) + 1
     except OSError:
         return 1
+
+
+def parse_rm_dirs(rm_dirs: Iterable[str]) -> Set[str]:
+    """Parse a list of possibly colon-separated dirs (or files or globs).
+    Return the set of all the dirs.
+
+    Used by cylc clean with the --rm option.
+    """
+    result: Set[str] = set()
+    for item in rm_dirs:
+        for part in item.split(':'):
+            part = part.strip()
+            if not part:
+                continue
+            is_dir = part.endswith('/')
+            part = os.path.normpath(part)
+            if os.path.isabs(part):
+                raise UserInputError("--rm option cannot take absolute paths")
+            if part == '.' or part.startswith('../'):
+                raise UserInputError(
+                    "--rm option cannot take paths that point to the "
+                    "run directory or above"
+                )
+            if is_dir:
+                # Preserve trailing slash to ensure it only matches dirs,
+                # not files, when globbing
+                part += '/'
+            result.add(part)
+    return result
+
+
+def glob_in_run_dir(run_dir: Union[Path, str], pattern: str) -> List[Path]:
+    """Execute a (recursive) glob search in the given run directory.
+
+    NOTE: this follows symlinks, so be careful what you do with the results.
+    """
+    return sorted(
+        Path(i) for i in glob.iglob(
```

Comment:
Strings and Pathlib `Path`s have different sort orders. We need the latter's behaviour of putting `name` before `name.ext` in case `name` is a dir and `name.ext` is a file.

Argh, just realised that `name.ext` could be a dir and `name` a file

Reply:
Ah ok.

> just realised that name.ext could be a dir

Yeah, that's pretty common, could provide a custom sort key to `sorted`.

Reply:
We need to order dirs before files because of https://github.com/cylc/cylc-flow/blob/9e3a30d287697c0ba221bdd16e6125565e0961e2/cylc/flow/workflow_files.py#L753-L755

which wouldn't work for the ASCII sort order:
```python
>>> sorted(['foo', 'foo/bar', 'foo.txt'])
['foo', 'foo.txt', 'foo/bar']
```
In this example, `foo/bar` is supposed to be skipped because we will already have deleted `foo`, but won't be as `skip_children_of = foo.txt` by the time we get to `foo/bar`

Reply:
Sort key's are functions which allow you to take control of the comparisons that determine the sort order:

```python
>>> sorted(['2', '01', '03'])
['01', '03', '2']
>>> sorted(['2', '01', '03'], key=lambda x: f'{int(x):02d}')
['01', '2', '03']
```

Reply:
> Strings and Pathlib `Path`s have different sort orders. We need the latter's behaviour of putting `name` before `name.ext` in case `name` is a dir and `name.ext` is a file.
> 
> Argh, just realised that `name.ext` could be a dir and `name` a file

This is resolved by latest commits